develop/plugin_tutorial.rst

```diff
@@ -158,402 +104,358 @@ special note of the section that looks like this::
 
-    Class = Random
+We also modify the docstring to document this function. Note the syntax
+on the first line: by convention, required arguments go in ``<>`` and optional
+arguments should be surrounded by ``[]``.
 
-And also some vim modeline stuff. Leave these as is, and we're finally done
-with plugin.py!
+The function body includes a new method
+:py:meth:`irc.replySuccess <supybot.callbacks.RichReplyMethods.replySuccess>`.
+This is just a generic "I succeeded" command which responds with whatever the
+bot owner has configured in ``config supybot.replies.success``.
+Also, by using ``@wrap``, we don't need to do any type checking inside the
+function itself - this is handled separately, and invalid argument values will
+cause the command to error before it reaches the wrapped function.
+
+With this alone you'd be able to make a range of useful plugin commands, but
+we'll go include some more examples to illustrate common patterns.
+
+Command handler with list-type arguments
+----------------------------------------
+The next sample command is named ``sample`` (no pun intended): it takes a random
+sample of arbitrary size from a list provided by the user:
+
+.. code-block::
+    :dedent: 0
+
+        @wrap(['int', many('anything')])
+        def sample(self, irc, msg, args, n, items):
+            """<number of items> <item1> [<item2> ...]
+
+            Returns a sample of the <number of items> taken from the remaining
+            arguments.  <number of items> must be less than the number
+            of arguments given.
+            """
+            if n > len(items):
+                # Calling irc.error with Raise=True is an alternative early return
+                irc.error('<number of items> must be less than the number '
+                          'of arguments.', Raise=True)
+            sample = self.rng.sample(items, n)
+            sample.sort()
+            irc.reply(utils.str.commaAndify(sample))
+
+The important thing to note is that list type arguments are rolled into one
+parameter in the command function by the ``many`` filter. Similar "multiplicity"
+handlers are documented :ref:`here <wrap-multiplicity-handlers>`.
+
+We also update the docstring to use the ``[]`` convention when surrounding
+optional arguments.
+
+For this function's body,
+:py:meth:`irc.error <supybot.callbacks.NestedCommandsIrcProxy.error>`
+is like
+:py:meth:`irc.replySuccess <supybot.callbacks.NestedCommandsIrcProxy.replySuccess>`
+but for error messages. We prefer using this instead of ``irc.reply`` for error
+signaling because its behaviour can be configured specially. For example, you
+can force all errors to go in private by setting the ``reply.error.inPrivate``
+option, and this can help reduce noise on a busy channel.
+Also, ``irc.error()`` with no text will return a generic error message
+configured in ``supybot.replies.error``, but this is not a valid call to
+:py:meth:`irc.reply <supybot.callbacks.NestedCommandsIrcProxy.reply>`.
```

Comment:
```suggestion
For this function's body,
:py:meth:`irc.error <supybot.callbacks.ReplyIrcProxy.error>`
is like
:py:meth:`irc.replySuccess <supybot.callbacks.ReplyIrcProxy.replySuccess>`
but for error messages. We prefer using this instead of ``irc.reply`` for error
signaling because its behaviour can be configured specially. For example, you
can force all errors to go in private by setting the ``reply.error.inPrivate``
option, and this can help reduce noise on a busy channel.
Also, ``irc.error()`` with no text will return a generic error message
configured in ``supybot.replies.error``, but this is not a valid call to
:py:meth:`irc.reply <supybot.callbacks.ReplyIrcProxy.reply>`.
```

ditto